### PR TITLE
Implement origin concept and browsing contextless documents.

### PR DESCRIPTION
**Pull request by emilio**
Rebase of #8658.

Just rebased and built it. I think there weren't more comments to be addressed, but if there are, let me know and I'll try to fix them.

cc @jdm

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9779)

### CODE REVIEW

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -2397,11 +2434,14 @@ impl DocumentMethods for Document {
 
     // https://html.spec.whatwg.org/multipage/#dom-document-cookie
     fn GetCookie(&self) -> Fallible<DOMString> {
-        // TODO: return empty string for cookie-averse Document
-        let url = self.url();
-        if !is_scheme_host_port_tuple(&url) {
+        if self.is_cookie_averse() {
+            return Ok(DOMString::from("".to_owned()));
```
Comment (KiChjang) [inline components/script/dom/document.rs]: `DOMString::new()` instead?

Comment (ecoal95) [inline components/script/dom/document.rs]: Done

**Comment by Ms2ger** [comment]
Yes, I still want to review this in detail.

**Comment by Ms2ger** [comment]
Pushed your rebase to the original PR.

**Pull request closed without merging**